lib/coordinator.js

```diff
@@ -138,33 +203,51 @@ function removeRoute (interceptor, url, ports, force = false) {
 
   const routes = interceptor[kRoutes]
   const roundRobin = routes.get(url)
-  let modified = false
+  if (!roundRobin) return
 
-  if (roundRobin) {
-    for (const port of ports) {
-      if (roundRobin.delete(port)) {
-        modified = true
-      }
+  const threadIds = []
+  for (const port of ports) {
+    if (roundRobin.ports?.includes(port)) {
+      threadIds.push(port[kThread])
     }
+  }
 
-    if (roundRobin.length === 0) {
-      routes.delete(url)
-    }
+  const promises = []
+  for (const threadId of threadIds) {
+    promises.push(removeThreadRoutes(interceptor, threadId))
   }
+  await Promise.all(promises)
+}
 
-  if (modified) {
-    const allPorts = new Set(
-      Array.from(routes.values())
-        .map(r => r.ports)
-        .flat()
-    )
+async function removeThreadRoutes (interceptor, threadId) {
+  const routes = interceptor[kRoutes]
+  const otherPorts = new Set()
 
-    for (const port of ports) {
-      notifyPortRemoval(port, allPorts, port[kThread])
+  for (const roundRobin of routes.values()) {
+    for (const port of roundRobin) {
+      if (port[kThread] === threadId) {
+        port[kOnReadyHook]?.()
+        port[kOnReadyHook] = null
+        continue
+      }
+      if (port[kWired]) {
+        otherPorts.add(port)
+      }
     }
   }
 
-  return modified
+  const promises = [commonRemoveRoute(interceptor, threadId)]
+
+  // Notify all ports in the roundRobin that the route has been removed
+  for (const otherPort of otherPorts) {
+    otherPort.postMessage({ type: MESSAGE_ROUTE_REMOVE, threadId })
+    promises.push(waitMessage(otherPort, m =>
+      m.type === MESSAGE_ROUTE_REMOVED &&
+      m.threadId === threadId
+    ))
+  }
+
+  await Promise.all(promises)
```

Comment:
Can we have a situation this never happens? Should we add a timer?

Reply:
I added a timeout.